app/src/behaviors/behavior_tap_hold.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (c) 2020 Cody McGinnis, Okke Formsma
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_tap_hold
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/endpoints.h>
+#include <zmk/event-manager.h>
+#include <zmk/events/position-state-changed.h>
+#include <zmk/events/keycode-state-changed.h>
+#include <zmk/events/modifiers-state-changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_NODE_EXISTS(DT_DRV_INST(0))
+
+/************************************************************  DATA SETUP */
+#define ZMK_BHV_TAP_HOLD_MAX_HELD 10
+#define ZMK_BHV_TAP_HOLD_MAX_CAPTURED_KC 40
+
+#define TH_TYPE_MOD_PREFERRED 0
+#define TH_TYPE_BALANCED 1
+#define TH_TYPE_TAP_PREFERRED 2
+
+
+// increase if you have keyboard with more keys.
+#define TH_POSITION_NOT_USED 9999 
+
+struct behavior_tap_hold_behaviors {
+  struct zmk_behavior_binding tap;
+  struct zmk_behavior_binding hold;
+};
+
+typedef k_timeout_t (*timer_func)();
+
+struct behavior_tap_hold_config {
+  timer_func tapping_term_ms;
+  struct behavior_tap_hold_behaviors* behaviors;
+  char* flavor;
+};
+
+// this data is specific for each tap-hold
+struct active_tap_hold {
+  s32_t position;
+  bool is_decided;
+  bool is_hold;
+  const struct behavior_tap_hold_config *config;
+  struct k_delayed_work work;
+  bool work_is_cancelled;
+};
+
+struct active_tap_hold* undecided_tap_hold = NULL;
+struct active_tap_hold active_tap_holds[ZMK_BHV_TAP_HOLD_MAX_HELD] = {};
+struct position_state_changed* captured_position_events[ZMK_BHV_TAP_HOLD_MAX_CAPTURED_KC] = {};
```

Comment:
it's a little weird to use global variables here, but it's the only option I saw. The listener to key-position-events does not know for which device it was called. We need the globals to select the correct `active_tap_hold` struct.